### ISSUE TITLE
horovod is summing the gradient or averaging gradient from all the workers?

### ISSUE THREAD

**Issue by ustcyue** [horovod/horovod#843]
This issue https://github.com/horovod/horovod/issues/390 mentioned that in the benchmark, gradients are summed up instead of being averaged. 

My question is, what is the default behavior for horovod? Looking at the source code https://github.com/horovod/horovod/blob/master/horovod/tensorflow/__init__.py#L230, it seems the default behavior is summing gradient from all workers? Is this the truth and is there anyway to control this behavior? Thanks!

**Comment by alsrgv** [comment]
@ustcyue, the default is to average: https://github.com/horovod/horovod/blob/master/horovod/tensorflow/__init__.py#L46

**Comment by ustcyue** [comment]
@alsrgv Thank you for pointing it out. Seems the only way to change this behavior is modifying the source code?

**Comment by tgaddair** [comment]
Hey @ustcyue, you can run `allreduce(tensor, average=False)` or create your own [DistributedOptimizer](https://github.com/horovod/horovod/blob/master/horovod/tensorflow/__init__.py#L151) implementation that does gradient summing.  No need to modify the source code.

**Comment by ustcyue** [comment]
Thank you @tgaddair , may I know why horovod uses average as the default? Is there any benchmark result for sum vs avg gradient?

**Comment by tgaddair** [comment]
They should converge the same in most cases, provided you scale the learning rate with the number of workers when averaging.  Facebook published a paper, [Accurate, Large Minibatch SGD: Training ImageNet in 1 Hour](https://arxiv.org/pdf/1706.02677.pdf), that talks about the idea of scaling the learning rate.

An advantage of averaging is gradient magnitudes that are invariant to the number of workers (and thus easier to reason about when experimenting).  Some regularization strategies like L2 regularization will penalize large weights, so averaging helps to prevent that from happening.

**Comment by ustcyue** [comment]
@tgaddair Got it, thank you for the explanation, it is very helpful. I will close the issue.

**Comment by pescap** [comment]
Update: Current version of horovod is without averaging, as opposed to the 2019 version, am I right? 

Now: 

https://github.com/horovod/horovod/blob/1fbdab55f0477fe8c21a80ef31cbe9614bd452b4/horovod/tensorflow/__init__.py#L58

Before:

https://github.com/horovod/horovod/blob/6643d6aa4abd583e5d254e2658e413a309f47a54/horovod/tensorflow/__init__.py#L46

**Comment by maxhgerlach** [comment]
That's not right, please read the doc string regarding `average` (deprecated) and `op` arguments.

**Comment by pescap** [comment]
Thank you for your quick reply!

https://github.com/horovod/horovod/blob/1fbdab55f0477fe8c21a80ef31cbe9614bd452b4/horovod/tensorflow/__init__.py#L85-L86

As you state, the default allreduce is with averaging.
